@brandond thanks for the quick review.  

This is a flaky issue in real code, several things need to roll low simultaneously for it to happen.
Here are tests to explain my point and reproduce data races caused by the absence of lock, please take a look.

`go.mod`

<details>

```
module cidrace

go 1.26.4

// https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/go.mod#L143
require github.com/yl2chen/cidranger v1.0.2
```

</details>

`race_test.go`

<details>

```
package cidrace

import (
	"net"
	"sync"
	"testing"

	"github.com/yl2chen/cidranger"
)

type rangerEntry struct{ network net.IPNet }

func (e rangerEntry) Network() net.IPNet { return e.network }

// go test -race
func TestPCTrieRangerConcurrentReadWrite(t *testing.T) {
    // https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L37
    ranger := cidranger.NewPCTrieRanger()
	_, network, _ := net.ParseCIDR("10.42.0.0/24")
	ip := net.ParseIP("10.42.0.5")

	const iterations = 100000
	var wg sync.WaitGroup
	wg.Add(3)

	// immitate t.Lock() of both writers (onChangeNode/onChangePod)
	// https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L120
	// https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L145
	var writeLock sync.Mutex

	// put pressure, immitate nodes/pods joining and leaving (onChangeNode/onChangePod)
	inserter := func() {
		defer wg.Done()
		for i := 0; i < iterations; i++ {
			writeLock.Lock()
			// https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L133
			ranger.Insert(rangerEntry{*network})
			writeLock.Unlock()
		}
	}

	remover := func() {
		defer wg.Done()
		for i := 0; i < iterations; i++ {
			writeLock.Lock()
			// https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L129
			ranger.Remove(*network)
			writeLock.Unlock()
		}
	}

	// reader, like dialBackend ContainingNetworks (no lock):
	reader := func() {
		defer wg.Done()
		for i := 0; i < iterations; i++ {
			// https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L217
			ranger.ContainingNetworks(ip)
		}
	}

	go inserter()
	go remover()
	go reader()
	wg.Wait()
}
```

</details>

`egress_map_test.go`

<details>

```
package cidrace

import (
	"sync"
	"testing"
)

// go test -race
func TestEgressMapConcurrentReadWrite(t *testing.T) {
    // https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L40
	egress := map[string]bool{}

	const iterations = 100000
	var wg sync.WaitGroup
	wg.Add(2)

	// writer, like onChangeNode:
	writer := func() {
		defer wg.Done()
		for i := 0; i < iterations; i++ {
			// https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L122
			egress["node-a"] = true
		}
	}

	// reader, like dialBackend:
	reader := func() {
		defer wg.Done()
		for i := 0; i < iterations; i++ {
			// https://github.com/k3s-io/k3s/blob/79b2e9ee84df391eb9722c874ebf231c5a10e0fc/pkg/daemons/control/tunnel.go#L224
			_ = egress["node-a"]
		}
	}

	go writer()
	go reader()
	wg.Wait()
}
```
</details>

Reproduce:
```
# map
go test -v -run TestEgressMapConcurrentReadWrite -count=1 .

fatal error: concurrent map read and map write
```

```
# map with race
go test -race -v -run TestEgressMapConcurrentReadWrite -count=1 .

WARNING: DATA RACE
Read at 0x00c000116aa8 by goroutine 10:
  cidrace.TestEgressMapConcurrentReadWrite.func2()        // reader
      egress_map_test.go:30

Previous write at 0x00c000116aa8 by goroutine 9:
  cidrace.TestEgressMapConcurrentReadWrite.func1()        // writer
      egress_map_test.go:21
--- FAIL: TestEgressMapConcurrentReadWrite
```
```
# trie with race
go test -race -v -run TestPCTrieRangerConcurrentReadWrite -count=1 .

WARNING: DATA RACE
Write at 0x00c00011e300 by goroutine 9:
  github.com/yl2chen/cidranger.(*prefixTrie).appendTrie()
  github.com/yl2chen/cidranger.(*prefixTrie).insert()
  github.com/yl2chen/cidranger.(*prefixTrie).Insert()
  cidrace.TestPCTrieRangerConcurrentReadWrite.func1()        // inserter

Previous read at 0x00c00011e300 by goroutine 11:
  github.com/yl2chen/cidranger.(*prefixTrie).containingNetworks()
  github.com/yl2chen/cidranger.(*prefixTrie).ContainingNetworks()
  cidrace.TestPCTrieRangerConcurrentReadWrite.func3()        // reader
--- FAIL: TestPCTrieRangerConcurrentReadWrite
```

--- a/pkg/daemons/control/tunnel.go
+++ b/pkg/daemons/control/tunnel.go
@@ -211,10 +211,10 @@ func (t *TunnelServer) dialBackend(ctx context.Context, addr string) (net.Conn, 
 
 	var nodeName string
 	var toKubelet, useTunnel bool
-	t.RLock()
 	if ip := net.ParseIP(host); ip != nil {
 		// Destination is an IP address, which could be either a pod, or node by IP.
 		// We can only use the tunnel for egress to pods if the agent supports it.
+		t.RLock()
 		if nets, err := t.cidrs.ContainingNetworks(ip); err == nil && len(nets) > 0 {
 			if n, ok := nets[0].(*tunnelEntry); ok {
 				nodeName = n.nodeName
@@ -228,13 +228,13 @@ func (t *TunnelServer) dialBackend(ctx context.Context, addr string) (net.Conn, 
 				logrus.Debugf("Tunnel server egress proxy CIDR lookup returned unknown type for address %s", ip)
 			}
 		}
+		t.RUnlock()
 	} else {
 		// Destination is a node by name, it is safe to use the tunnel.
 		nodeName = host
 		toKubelet = true
 		useTunnel = true
 	}
-	t.RUnlock()
 
 	// If connecting to something hosted by the local node, don't tunnel
 	if nodeName == t.config.ServerNodeName {

